### ISSUE TITLE
Task Group visualization doesn't scale to thousands of task groups

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#5021]
I just saw an example where there were a thousand TaskGroups, each of size one.  This is arguably the fault of the code that produced such a wonky setup, but we should be robust to it.  

As a solution, I propose that if there are more than, say, 200 TaskGroups, that we only display those that have many tasks

Actually, that doesn't quite work for a few reasons:

1.  We may lose important dependencies like `sum()` 
2.  That may break parts of the graph, resulting in odd holes, killing our layout algorithm

But!  Maybe we can be clever here in some way!  Thoughts welcome.  cc @ncclementi @jrbourbeau 

In the specific case that I saw the thousand single-task groups were all data creation tasks without dependencies.  @ian-r-rose may be able to provide a reproducer.

**Comment by jrbourbeau** [comment]
Hmm thanks for bringing this up @mrocklin

> In the specific case that I saw the thousand single-task groups were all data creation tasks without dependencies. @ian-r-rose may be able to provide a reproducer.

Here's a contrived example which simulates this (though a real-world reproducer is certainly welcome)

```python
from distributed import Client

client = Client()
fs = [client.submit(lambda x: x, i, key=i) for i in range(200)]
```

**Comment by mrocklin** [comment]
So, in principle I think that we would be comfortable merging TaskGroups if following conditions are met

1.  There are relatively few tasks in the group (the most common failure case is 1) (although maybe not this one, xarray breaks this assumption)
2. The prefix among these many groups are the same `len(set(tg.prefix for tg in groups)) == 1`
3. The dependencies among these many groups are the same `len(set(sorted(tg.dependencies) for tg in groups)) == 1`
4. The dependents among these many groups are the same `...`

We could imagine doing this in the visualization section, maybe by hashing things down 

```python
for prefix in self.scheduler.prefixes:
    if len(prefix.groups) > 10:
        def f(tg):
            return hash((tg.prefix, tuple(tg.dependencies), tuple(tg.dependents)))

        groups_of_groups = toolz.groupby(f, prefix.groups)
        ... do something clever with each group
```

However it would be better to solve this on the scheduler itself, rather than in just this visualization.  Solving it at the scheduler would be more general, and would apply to other places where these are used.

So we might try to deduplicate task groups as they are generated in `Scheduler.update_graph`.  This would probably have to happen after the `ts.add_dependency` call in order to make sure that we had dependency information.  We might then look at all prefixes that were touched during the update_graph call, and then for all of those prefixes that had several groups all with the same dependency/dependent information we might choose to merge those groups into one (we would pick one, remove the others, and set the `ts._group` attribute to all relevant tasks).

It might make more sense not to generate the TaskGroup as we create a new task, but later in update_graph instead, after all TaskStates have been created and their dependencies set.

**Comment by mrocklin** [comment]
I tried to generate this situation in the wild with dask.delayed and Client.submit but failed (everything worked smoothly) as well as `dask_ml.datasets.make_classification` (this came up in a client example).  @ian-r-rose and @gjoseph92 you both mentioned that you had run into this situation.  If it's easy, can I ask you to share a small reproducer?

**Comment by ian-r-rose** [comment]
Yeah, @mrocklin the issue we saw wasn't due to `make_classification`, but the concatenation of many similar dataframes. In this case it was from repeated resampling of some root dataset. So each sub-dataframe has all the same dependencies and also the same dependents (and prefixes), but are treated as different task groups.

```python
import dask.dataframe as dd
import distributed
from dask.datasets import timeseries

ddf = timeseries(freq="1H").repartition(npartitions=1)
ddf =dd.concat([
    ddf.sample(frac=0.1) for _ in range(100)
])
client = distributed.Client()
distributed.wait(ddf.persist())
```

This does in fact meet the criteria you suggest above, and I agree that it seems like a reasonable way forward.

**Comment by ian-r-rose** [comment]
A quick experiment with combining similar task groups (just looking at prefixes and dependencies at the moment) within the TaskGroupGraph visualization: https://github.com/ian-r-rose/distributed/commit/ab6d29b565687ba3d8d5de223e4876678a98f70d
